File: samples/vdsp_op/custom_op/argmax/argmax_prof.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../../..")
sys.path.append(common_path)

from argmax_op import ArgmaxOp
from common.model_profiler import ModelProfiler
from easydict import EasyDict as edict
import argparse
import ast


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--elf_file",
        default="/opt/vastai/vaststreamx/data/elf/planar_argmax",
        help="elf file",
    )
    parser.add_argument(
        "-d",
        "--device_ids",
        default="[0]",
        help="device ids to run",
    )
    parser.add_argument(
        "-i",
        "--instance",
        default=1,
        type=int,
        help="instance number for each device",
    )
    parser.add_argument(
        "-s",
        "--shape",
        help="model input shape",
    )
    parser.add_argument(
        "--iterations",
        default=10240,
        type=int,
        help="iterations count for one profiling",
    )
    parser.add_argument(
        "--percentiles",
        default="[50, 90, 95, 99]",
        help="percentiles of latency",
    )
    parser.add_argument(
        "--input_host",
        default=0,
        type=int,
        help="cache input data into host memory",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    elf_file = args.elf_file
    device_ids = ast.literal_eval(args.device_ids)
    batch_size = 1
    instance = args.instance
    iterations = args.iterations
    queue_size = 0
    input_host = args.input_host
    percentiles = ast.literal_eval(args.percentiles)
    shape = ast.literal_eval(args.shape)

    ops = []
    contexts = []
    for i in range(instance):
        device_id = device_ids[i % len(device_ids)]
        op = ArgmaxOp("planar_argmax_op", elf_file, device_id)
        ops.append(op)
        if input_host:
            contexts.append("CPU")
        else:
            contexts.append("VACC")
    config = edict(
        {
            "instance": instance,
            "iterations": iterations,
            "batch_size": batch_size,
            "data_type": "float16",
            "device_ids": device_ids,
            "contexts": contexts,
            "input_shape": shape,
            "percentiles": percentiles,
            "queue_size": queue_size,
        }
    )

    profiler = ModelProfiler(config, ops)
    print(profiler.profiling())

File: samples/vdsp_op/custom_op/argmax/argmax_prof.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <chrono>
#include <thread>

#include "argmax_op.hpp"
#include "common/cmdline.hpp"
#include "common/model_profiler.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("elf_file", '\0', "elf_file path", false,
                        "/opt/vastai/vaststreamx/data/elf/planar_argmax");
  args.add<std::string>("device_ids", 'd', "device id to run", false, "[0]");
  args.add<std::string>("shape", '\0', "input shape [c,h,w]", false,
                        "[19,512,512]");
  args.add<uint32_t>("instance", 'i',
                     "instance number or range for each device", false, 1);
  args.add<int>("iterations", '\0', "iterations count for one profiling", false,
                10240);
  args.add<std::string>("percentiles", '\0', "percentiles of latency", false,
                        "[50,90,95,99]");
  args.add<bool>("input_host", '\0', "cache input data into host memory", false,
                 0);
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  auto elf_file = args.get<std::string>("elf_file");
  auto device_ids = vsx::ParseVecUint(args.get<std::string>("device_ids"));
  uint32_t batch_size = 1;
  auto instance = args.get<uint32_t>("instance");
  auto iterations = args.get<int>("iterations");
  auto input_host = args.get<bool>("input_host");
  uint32_t queue_size = 0;
  auto percentiles = vsx::ParseVecUint(args.get<std::string>("percentiles"));
  auto shape = vsx::ParseShape(args.get<std::string>("shape"));

  std::vector<std::shared_ptr<vsx::ArgmaxOp>> ops;
  ops.reserve(instance);
  std::vector<vsx::Context> contexts;
  for (uint32_t i = 0; i < instance; i++) {
    uint32_t device_id = device_ids[i % (device_ids.size())];
    if (input_host) {
      contexts.push_back(vsx::Context::CPU());
    } else {
      contexts.push_back(vsx::Context::VACC(device_id));
    }
    ops.push_back(std::make_shared<vsx::ArgmaxOp>("planar_argmax_op", elf_file,
                                                  device_id));
  }
  std::cout << "shape:" << shape << ", instance:" << instance
            << ", iterations:" << iterations << ",batch_size:" << batch_size
            << ", queue_size: " << queue_size << std::endl;
  vsx::ProfilerConfig config = {instance,      iterations,  batch_size,
                                vsx::kFloat16, device_ids,  contexts,
                                {shape},       percentiles, queue_size};
  vsx::ModelProfiler<vsx::ArgmaxOp> profiler(config, ops);
  std::cout << profiler.Profiling() << std::endl;
  return 0;
}
